Prep 5.0.0-M1

matrei opened the pull request:
This PR preps the project for a 5.0.0-M1 release. Please refer to the individual commit messages for detailed descriptions of the changes.

matrei committed:
fix(deps): set Grails version `7.0.0-M1`

matrei committed:
fix(deps): upgrade `greenmail` to `2.1.2`

- Upgrade from `2.1.0` to `2.1.2`.
- [2.1.2 Release Notes](https://github.com/greenmail-mail-test/greenmail/releases/tag/release-2.1.2).
- [2.1.1 Release Notes](https://github.com/greenmail-mail-test/greenmail/releases/tag/release-2.1.1).

matrei committed:
chore: add `.sdkmanrc`

- Usage: https://sdkman.io/usage/#env-command

matrei committed:
build(deps): upgrade Gradle wrapper to `8.12`

matrei committed:
fix: add `@CompileStatic` to `Application`

matrei committed:
build(deps): upgrade `asciidoctor` Gradle plugin to `4.0.4`

- Upgrade from `4.0.2` to `4.0.4`
- [4.0.4 Release Notes](https://github.com/asciidoctor/asciidoctor-gradle-plugin/releases/tag/release_4_0_4)
- [4.0.3 Release Notes](https://github.com/asciidoctor/asciidoctor-gradle-plugin/releases/tag/release_4_0_3)

matrei committed:
build: change repository order

- Move `repo.grails.org` to the top

matrei committed:
build: rename `publishing.gradle` file

- Rename `publishing.gradle` to `publish-config.gradle` file to be in-line with the other include files.

matrei committed:
fix: update `MailGrailsPlugin`

- New compatibility version: Grails 7.
- Update with new repo and organization home.

matrei committed:
build: update `publish-config.gradle`

- Update with new GitHub organization home.

matrei committed:
build: remove `buildSrc` and version catalog

- Consensus in Grails team is to remove `buildSrc` (where possible) and version catalogs for simplicity and consistency across repositories.

matrei committed:
build: switch to `grails-publish`

- `grails-publish` Gradle Plugin is a new way to simplify publishing in Grails 7.

matrei committed:
ci: pin GitHub workflows to `ubuntu-24.04`

- Prevents the runner from being a moving target by using a fixed version.

matrei committed:
chore(ci): place `java-version` first in `setup-java` action

- Aligns with the typical property order for actions/setup-java for consistency.

matrei committed:
build: re-order dependencies by scope

- Dependencies are now arranged in scope order for better readability and consistency.

matrei committed:
build: redo build composition

- `grails-plugin` is now handling disabling Boot tasks
- Refactor `java-config.gradle` and `grails-plugin-config.gradle` into the main build file as these are so small it is not worth having them extracted in a single-project build.
- Extract java version to `gradle.properties` and use `compileJava.options.release` instead of `java.sourceCompatibility`.
- Extract version of `asciidoctor` gradle plugin to `gradle.properties`.
- Align Gradle plugin inclusion to use apply, as Grails plugins must use that.

matrei committed:
build: remove unused properties

- Removed `isReleaseVersion` and `isSnapshot` properties, which are no longer needed after migrating to `grailsPublish`.

matrei committed:
build(deps): update Gradle plugins for build scans

matrei committed:
build: add tags to the build scans

sbglasius reviewed (approved):
Looks good to me. Still releasing to `org.grails.plugins` group?

matrei commented:
> Looks good to me. Still releasing to org.grails.plugins group?

@sbglasius Is this not correct?

matrei committed:
build: add `buildSrc` for `asciidoctor` plugin

Asciidoctor Gradle Plugin works smoother with when included in `buildSrc`.

matrei committed:
build: update `groovydoc` settings

matrei committed:
build: extract integration tests to separate app

matrei committed:
build: update test logging events

matrei committed:
build: add test dependency `junit-platform-launcher`

Required by Gradle 9

matrei committed:
docs: improve documentation generation process

- Replace `dependsOn` with `inputs` and `outputs` in Gradle tasks.
- Merge documentation into a single file:
  - Removed unnecessary includes due to the small documentation size.
  - Ensure correct "last updated" date in footer, even if
    the index page isn't updated (see
    https://github.com/asciidoctor/asciidoctor/issues/1163).

matrei committed:
fix!: update Maven coordinates

Change maven coordinates from `org.grails.plugins:mail` to
`io.github.gpc:grails-mail`.

matrei committed:
fix(deps): update `develocity` gradle plugin

Update from `3.19` to `3.19.2`.

matrei committed:
fix(deps): update `grailsVersion` to `7.0.0-M3`

matrei committed:
chore: cleanup

matrei committed:
build: clarify dependency usage and scope

matrei committed:
refactor: Use Spring Boot `@AutoConfiguration`

matrei committed:
refactor: move classes out of `grails-app`

Simplify the project by keeping all sources in `src/main/groovy`.

matrei committed:
chore: cleanup

matrei committed:
test: update tests

- Add functional test
- Refactor tests

matrei committed:
build: update documentation generation

- Rename to shorter file name
- Output groovydoc into `docs/api`
- DRY up `docs` task

matrei committed:
test: add taglib tests

matrei committed:
fix: set `PlainTextMailTagLib` as provided artefact

This is necessary for application to find the tag library as it has
been defined in `src/main/groovy` and not in `grails-app/taglib`.

matrei committed:
ci: update CI

- Publish Snapshots to GitHub packages
- Use Apache Grails GitHub Actions

matrei committed:
ci: set permissions at job level

As setting on step level is not supported.

matrei committed:
ci: set correct permission attributes

matrei committed:
build: update java version in `.sdkmanrc`

matrei committed:
build: update gradle wrapper version to 8.13

matrei committed:
ci: fix typo in `release.yml`

matrei committed:
ci: remove obsolete permissions block `release-notes.yml`

matrei committed:
ci: update build scan plugin versions and tags in testapp1

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -32,10 +28,14 @@ jobs:
           develocity-access-key: ${{ secrets.DEVELOCITY_ACCESS_KEY }}
       - name: "🏃 Run tests"
         run: ./gradlew check
+      - name: "🏃 Run integration tests"
```
Doesn't check run the integration tests?

matrei replied:
No, the integration tests are in a decoupled test app, that uses includeBuild.

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -32,10 +28,14 @@ jobs:
+        working-directory: ./examples/testapp1
+        run: ./gradlew integrationTest
   publish_snapshot:
+    if: github.event_name == 'push'
```
Don't you want it to publish the snapshot if it's manually run too?  i.e. add workflow_dispatch to the top of this file & adjust this event check

matrei replied:
Hmm, I have never had that need. Will add it.

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -50,23 +50,23 @@ jobs:
-          MAVEN_PUBLISH_URL: ${{ secrets.MAVEN_PUBLISH_SNAPSHOT_URL }}
+          MAVEN_PUBLISH_USERNAME: ${{ secrets.GITHUB_ACTOR }}
+          MAVEN_PUBLISH_PASSWORD: ${{ secrets.GITHUB_TOKEN }}
+          MAVEN_PUBLISH_URL: 'https://maven.pkg.github.com/gpc/grails-mail'
```
I've never used the packages for snapshots.  The github docs say this is supported, will be curious how this works.

matrei replied:
Let's try it out! (Unless we decide to move the repo to github/grails-plugins)

matrei replied:
As we moved to `grails-plugins` I have changed snapshot publishing to https://repo.grails.org/artifactory/plugins3-snapshots-local

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -50,23 +50,23 @@ jobs:
+        permissions:
+          packages: write # publishes snapshot to GitHub Packages
         env:
+          GITHUB_MAVEN_PASSWORD: ${{ secrets.GITHUB_TOKEN }}
```
The only reason we needed GITHUB_MAVEN_PASSWORD is if we wanted to access a github maven repo, I don't see this defined in any of the repo blocks in the gradle files - I would drop this.

jdaugherty commented on .github/workflows/gradle.yml:
```diff
@@ -50,23 +50,23 @@ jobs:
-        uses: grails/github-pages-deploy-action@grails
+        uses: apache/grails-github-actions/deploy-github-pages@asf
+        permissions:
+          content: write # updates gh-pages branch
```
Wait, you can do this at the step level?!

matrei replied:
No, that was a mistake. It has been changed to job level.

jdaugherty commented on .github/workflows/release.yml:
```diff
@@ -2,12 +2,14 @@ name: "Release"
     runs-on: ubuntu-24.04
+    permissions:
+      packages: read  # for pre-release workflow
+      contents: write  # to create release
```
since the pre-release workflow is being used, it will update the version if in gradle.properties and update the "tag" of the release.  Additionally, the docs are committed in the doc publish.

I would suggest updating your comment to something like:

      # to commit changes related to the release and publish documentation to gh-pages

matrei replied:
Thanks, that was a copy/paste!

jdaugherty commented on .github/workflows/release.yml:
```diff
@@ -2,12 +2,14 @@ name: "Release"
+    permissions:
+      packages: read  # for pre-release workflow
+      contents: write  # to create release
+      issues: write  # to modify milestones
```
I can't recall if I kept the milestone support in the files, if I did, I likely made it so that it silently fails if it doesn't exist or has a problem.  We should probably test this.

jdaugherty commented on .github/workflows/release.yml:
```diff
@@ -21,16 +23,16 @@ jobs:
       - name: "📝 Store the current release version"
-        id: release_version
-        run: echo "release_version=${GITHUB_REF:11}" >> $GITHUB_OUTPUT
+      - run: |
```
I think I had to do this somewhere else too.  The pre-release script could be changed to export this value for us.  We could then eliminate this step in several builds.  A better name for pre-release may even be "set-release-version"

jdaugherty commented on .github/workflows/release.yml:
```diff
@@ -21,16 +23,16 @@ jobs:
       - name: "📤 Publish artifacts to Sonatype"
-        id: publish_to_sonatype
         env:
           GITHUB_MAVEN_PASSWORD: ${{ secrets.GITHUB_TOKEN }}
```
GITHUB_MAVEN_PASSWORD is likely unneeded per other comment

jdaugherty commented on build.gradle:
```diff
@@ -27,38 +15,78 @@ dependencies {
-    // to the compileClasspath of plugin users.
-    implementation 'org.grails:grails-gsp'
-    implementation 'org.grails:grails-web-common'
+    api 'org.grails:grails-gsp', {
```
Does the mail plugin really require the gsp plugin? This was implementation before and some people may prefer sending only text?

matrei replied:
The classes listed are exposed in the public api of the plugin.

matrei replied:
Also, grails-gsp will still be included in the app with implementation, only it will not be on the compile classpath.

jdaugherty replied:
Before it was implementation though so if you didn't use the gsp functions it would work, no?

jdaugherty replied:
depends if you include transitive=false when importing, but I guess this is ok.

jdaugherty commented on build.gradle:
```diff
@@ -27,38 +15,78 @@ dependencies {
+    api 'org.grails:grails-gsp', {
+        // api: GroovyPageTemplate, GroovyPagesTemplateEngine
+    }
+    api 'org.grails:grails-web-common', {
```
I don't think you want to expose web-common - mail can be used on the backend with no front end needed.

matrei replied:
`GroovyPagesUriService` is exposed in public api, and the other classes are used in implementation. This makes `grails-web-common` a direct dependency.

jdaugherty commented on build.gradle:
```diff
@@ -27,38 +15,78 @@ dependencies {
-
-    integrationTestRuntimeOnly 'org.springframework.boot:spring-boot-starter-tomcat'
-    integrationTestRuntimeOnly 'org.grails:grails-web-testing-support'
+    testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
```
Why was web-testing-support removed if you're including gsp?  Shouldn't that be required?

matrei replied:
I don't see why `web-testing-support` should be required. Can you elaborate?

jdaugherty replied:
Isn't web testing support used for mocking tag libs, etc?  Does this project only have integration tests?  I would have assumed we would leave it for unit tests if people are inclined to test functions in isolation.

jdaugherty commented on buildSrc/build.gradle:
```diff
@@ -0,0 +1,20 @@
+    id 'groovy-gradle-plugin'
+}
+
+file('../gradle.properties').withInputStream {
```
One downside of doing this is environment variables don't override project properties.

matrei replied:
Good catch! I'll leave it for now and deal with it if it becomes a problem.

jdaugherty commented on src/docs/1. Introduction.adoc:
Am I missing something?  Where are the docs now?

matrei replied:
They are all aggregated in `index.adoc`.

jdaugherty commented on src/main/groovy/grails/plugins/mail/MailService.groovy:
```diff
@@ -57,18 +66,20 @@ class MailService implements InitializingBean, DisposableBean {
 
-	MailMessage sendMail(Config config, @DelegatesTo(strategy = Closure.DELEGATE_FIRST, value = MailMessageBuilder) Closure callable) {
-		return sendMail(toMailProperties(config), callable)
+	MailMessage sendMail(
```
Is there an extra new line before Config config?

matrei replied:
As the second parameter is so long, I chose to have parameters on new lines.

jdaugherty replied:
It looked like there was a double new line here after the function.  I think it's just github's diff render.  Thanks for the clarification.

jdaugherty commented on src/main/groovy/grails/plugins/mail/SendMail.groovy:
```diff
@@ -17,17 +17,15 @@ package grails.plugins.mail
 
 @CompileStatic
-@Enhances([DefaultGrailsServiceClass.SERVICE, DefaultGrailsControllerClass.CONTROLLER])
+@Enhances(['Controller', 'Service'])
```
Why remove the variables?

matrei replied:
I think they they are unlikely to change and easier to read and understand.

jdaugherty commented on build.gradle:
```diff
@@ -27,38 +15,78 @@ dependencies {
-
-    integrationTestRuntimeOnly 'org.springframework.boot:spring-boot-starter-tomcat'
-    integrationTestRuntimeOnly 'org.grails:grails-web-testing-support'
+    testRuntimeOnly 'org.junit.platform:junit-platform-launcher'
```
Do you intend to never write any integration tests in the plugin now?  (web-testing-support is missing)

matrei replied:
At the moment there no integration tests (they are moved to testapp1) so it's not needed.

jdaugherty commented on examples/testapp1/grails-app/conf/spring/resources.groovy:
```diff
@@ -0,0 +1,3 @@
+// Place your Spring DSL code here
+beans = {
```
Why even define an empty resources.groovy?

jdaugherty commented on examples/testapp1/grails-app/init/testapp1/BootStrap.groovy:
```diff
@@ -0,0 +1,10 @@
+    GreenMailService greenMailService
+
+    def init = {
+        greenMailService.start()
```
You should add a stop on the destroy

matrei replied:
It's stopped in `GreenMailService` with `@PreDestroy`. That way it should always be stopped. I'm not 100% it is necessary to do it this way.

matrei commented:
> If we stick with `org.grails.plugins`, grails-mail could be moved to https://github.com/grails-plugins/ like https://github.com/grails-plugins/grails-audit-logging-plugin and https://github.com/grails-plugins/grails-web-console.
> 
> https://github.com/gpc/ is trying to use groupid: `io.github.gpc`

@jamesfredley We, could move it again :smile:. I think @sbglasius also advocated for that in the Grails Planning Meeting on 2024-10-17.

matrei committed:
test: remove used `resources.groovy`

sbglasius commented:
In general, all non-core plugins under the GPC could move to use org.grails.plugins?

matrei committed:
chore: cleanup tab -> spaces

There was a mix of these so this commit aligns
the code on using spaces.

matrei committed:
ci: pr feedback

Remove unnecessary environment variable.

matrei committed:
ci: pr feedback

Add workflow dispatch to be able to publish snapshots
manually.

matrei committed:
ci: pr feedback

matrei committed:
build(testapp1): allow overriding version properties

This commit ensures that Gradle properties are only inherited
from the main project if they are not already set, for example
using `-P`.

Additionally, only properties whose names end with `Version`
will be inherited.

matrei committed:
chore: clarify test dependency usage

matrei committed:
chore: prepare for move to `grails-plugins` GitHub org

matrei committed:
ci: set correct env var `GRAILS_PUBLISH_RELEASE`

matrei committed:
ci: publish snapshots to `repo.grails.org`

matrei committed:
build: switch to grails at apache coordinates

matrei committed:
build: update testapp1 to use grails at apache coordinates

matrei committed:
build: update develocity
